### ISSUE TITLE
Plot anomaly values on a log scale

### ISSUE THREAD

**Issue by rhyolight** [rhyolight/smartthings.htm.bridge#4]
http://dygraphs.com/tests/logscale.html

**Comment by jefffohl** [comment]
I may be missing something obvious here, but given that these values are between 0 and 1, how should this work? For a log scale to be useful, I would think the data would have to be something like this:

0.1
10
100
1000
etc..

**Comment by rhyolight** [comment]
[Does this help?](http://math.stackexchange.com/questions/57429/functions-similar-to-log-but-with-results-between-0-and-1)

**Comment by jefffohl** [comment]
Not really. My understanding is that most of the data points in the anomaly scores are within the same order of magnitude. That is, a typical series might look like this:

0.1, 0.2, 0.3, 0.4, 0.5, etc..

as opposed to this:

0.00001, 0.0002, 0.003, 0.04, etc..

Or am I wrong about that?

Another way of asking is - what is the visual effect you are looking for?

**Comment by rhyolight** [comment]
From our experience with the anomaly likelihoods generated for HTM for IT, the log plot was pretty useful. We did not flag an anomaly unless the anomaly likelihood was over 0.9999 or something like that, so it a big difference between 0.9, 0.99, 0.999, etc. So yes, I think you are wrong about the order of magnitude -- at least how we treat it.

**Comment by jefffohl** [comment]
OK - that makes sense.

**Comment by jefffohl** [comment]
Would a function like this be suitable?

```math
f(x) = 1 / -log(x)
```
I tried using the `logscale` setting DyGraphs, but it doesn't seem to like working with the value range of 0-1, and doesn't render anything. Or perhaps I am doing it wrong.

**Comment by jefffohl** [comment]
Actually, simply
```
f(x) = 1 / (1 - x)
```
should do the trick, yes?

A series such as,

```
0,
0.1,
0.2,
0.3,
0.4,
0.5,
0.6,
0.7,
0.8,
0.9,
0.95,
0.98,
0.99,
0.999,
0.9999,
0.99999
0.999999
```
would then be transformed into

```
1
1.111
1.25
1.428
1.666
2
2.5
3.333
5
10
20
50
100
1000
10000
100000
1000000
```
and then use logscale on the chart to render it.

**Comment by rhyolight** [comment]
I'm not sure, I would have to see the chart. If it looks good to you, create a PR and I'll test it out.

**Comment by jefffohl** [comment]
Actually, this works great - except when the initial value is 1 - then it comes out to Infinity, which of course can't be plotted.

**Comment by jefffohl** [comment]
This can be resolved by subtracting a number from each value in the series, so that no value is ever 1. So, changing the formula to this can work:

```javascript
function f(x) {
    var y = x - 0.0001;
    return 1 / (1 - y);
 }
````

As to whether this is visually useful is another question. Since most of the data coming out of the HITC server right now is somewhat spotty, I am going to use some made up data to show you what different transforms will look like. The data looks like this:

```javascript
[
    0,
    0.1,
    0.2,
    0.3,
    0.4,
    0.5,
    0.6,
    0.7,
    0.8,
    0.9,
    0.95,
    0.98,
    0.99,
    0.999,
    0.9999,
    0.99999,
    0.999999,
    1.0,
    0.99999,
    0.9999,
   // ... (repeats 5 times - for a total of 100 rows)
 ]
```

I have two transform functions to show, using this dummy data:

1:
```javascript
function f(x) {
    var y = x - 0.0001;
    return 1 / (1 - y);
}
```

This is what the plot looks like with this transform:

![fraction](https://cloud.githubusercontent.com/assets/641147/12097299/cf5a0a58-b2cf-11e5-9cfe-e4cc3ce9966f.png)

2: 
```javascript
function f(x) {
    return Math.log(x);
}
````
This is what the plot looks like using this transform:

![logscale](https://cloud.githubusercontent.com/assets/641147/12097311/e31286c4-b2cf-11e5-8bdd-af17596a4c6e.png)

Given these options I have some questions:

* Is the dummy data flawed?
* Which of these plots do you think are more useful?

**Comment by rhyolight** [comment]
I'm confused because on both of your example plots, the Y-axis is linear, not logarithmic. I was expecting something more like this:

<img width="515" alt="screen shot 2016-01-04 at 1 13 49 pm" src="https://cloud.githubusercontent.com/assets/15566/12098041/1b5d0ecc-b2e5-11e5-8815-5c586b161220.png">

I highlighted the bottom of the chart because that is the `0.0` - `1.0` portion, but you see what I mean by the scale of the axis being logarithmic not the data?

I think the confusion is because you are expecting to transform the value, but they should be left alone. The Y axis is what needs to change, so that the same values are plotted on a log scale.

**Comment by jefffohl** [comment]
Yes, I see what you mean, but I am not sure how to get DyGraphs to display a log scale on the Y axis where each increment is a fraction closer to 1. Usually, logscales are used for sets with very large numbers, right? We sort of have the opposite of that. Just setting the option `logscale : true` in the DyGraphs options doesn't work, I am assuming because all of the values are below 1. (log(1) = 0 and log(0.1) = -1)

It would be nice to get the chart to render like this:
```
|                                                     | 0.99999
|                                                     | 0.9999
|                                                     | 0.999
|                                                     | 0.99
|                                                     | 0.9
|                                                     | 0
-------------------------------------------------------
```

Not knowing how to do this without modifying DyGraphs source code, I was looking for a way to get the desired visual effect, which my understanding is: to make it easy to spot values over 0.9 and be able to discern the difference between subtle differences - such as between 0.999 and 0.9999.

I am assuming that you have not done this elsewhere using DyGraphs? It seems that you have taken this approach elsewhere. How did you deal with this problem there?

It's quite possible that I am simply confused due to being Bad At Math™, so if that is the case, feel free to advise.

**Comment by jefffohl** [comment]
One possible solution would be to write a plugin for DyGraphs. Looking into that.

**Comment by jefffohl** [comment]
Ack! OK, I found out the reason why it would not do logscale on the anomaly score series. It wasn't because of the data being between 0 and 1. It just isn't supported on the Y2 axis. This is an open issue on DyGraphs: https://github.com/danvk/dygraphs/issues/324

So, I guess we are stuck with either transforming the data beforehand, or making a contribution to the DyGraphs community.

**Comment by rhyolight** [comment]
Nice find. I suggest we put this on the back-burner for now, unless you are
interesting in diving into some Dygraphs code. I am not in favor of
transforming the underlying data.

---------
Matt Taylor
OS Community Flag-Bearer
Numenta

**Comment by jefffohl** [comment]
OK - backburner it for now. If this charting starts to get heavier use, we can consider contributing to DyGraphs, or possibly switching to D3 or something where we have more flexibility.
